### PR TITLE
InstanceShutdownByProviderID should return true when instance is stopped or terminated in aws

### LINKED ISSUE #79435: InstanceShutdownByProviderID should return true when instance is stopped or terminated in aws
**What happened**:
InstanceShutdownByProviderID aws impl should check if the instance is terminated. This will cause confusions:

1. The interface implies that it will return true if instance is "shutdown", i.e. the machine is not running, and a "terminated" ec2 instance will also fit in to this category
2. This will cause terminated node miss the shutdown taint, with only unreachable taint, daemonset will tolerate https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/util/daemonset_util.go#L49 , but adding a desired replica on terminated node does not make sense

I'm aware that #69409 removed the "terminated" check, and it result in leaving nodes hang but without proper tainting (i.e. shutdown taint). Node termination will result in the following behavior:

- node unreachable
- node marked as "shutdown" (add a shutdown taint, which no application is supposed to tolerate)
- node disappear in cloud provider (i.e. after an hour)
- v1.Node gets deleted from cloud provider

By leaving the v1.Node object hanging for an hour or so can assist debugging, and make the behavior consistent with cloud provider, i.e. https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html

**What you expected to happen**:
"Shutdown" means "terminated" or "stopped" in aws

**How to reproduce it (as minimally and precisely as possible)**:
Terminate an instance in aws, shutdown taint is not added to the node

/kind bug
/sig aws
/assign zhan849

### PR DESCRIPTION
**Pull request by zhan849**
**What type of PR is this?**
/kind bug
/sig aws

**What this PR does / why we need it**:
Instance shutdown means terminated OR stopped in aws, return true only when instance is stopped will cause confusions:
1. The interface implies that it will return true if instance is "shutdown", i.e. the machine is not running, and a "terminated" ec2 instance will also fit in to this category
2. This will cause terminated node miss the shutdown taint, with only unreachable taint, daemonset will tolerate https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/util/daemonset_util.go#L49 , but adding a desired replica on terminated node does not make sense 

This PR checks terminated state for instance as well and added test for `InstanceShutdownByProviderID()`

**Which issue(s) this PR fixes**:
Fixes #79435

**Special notes for your reviewer**:
/cc @justinsb @jsafrane @micahhausler @gnufied 
/assign @mcrute

**TL;DR:** this fix does not change current behavior that after node is shutdown (stopped or terminated), we keep the `v1.Node` object in api server until the instance disappear from cloud provider, but just make node lifecycle controller to add a shutdown taint to the `v1.Node` object which makes it less confusing to other apps such as daemonset.

I'm aware that https://github.com/kubernetes/kubernetes/issues/69409 removed the "terminated" check, but with proper tainting (i.e. shutdown taint), node termination will result in the following behavior:

- node unreachable
- node marked as "shutdown" (add a shutdown taint, which no application is supposed to tolerate)
- node disappear in cloud provider (i.e. after an hour)
- v1.Node gets deleted from cloud provider

By leaving the v1.Node object hanging for an hour or so can assist debugging, the current behavior is to let the node hang after it gets terminated, but without a proper shutdown taint, and this change just made node lifecycle controller to add a shutdown taint to the terminated node.

**Does this PR introduce a user-facing change?**:
```release-note
InstanceShutdownByProviderID should return true when instance is stopped or terminated in aws
```

### CODE REVIEW

**Comment by mcrute** [comment]
Thanks for the patch @zhan849!

/ok-to-test
/lgtm
/approve

**Comment by justinsb** [comment]
/hold

I think this might block deletion of terminated nodes.... putting a hold on while I check

**Comment by justinsb** [comment]
Yes, I think this blocks deletion of terminated nodes: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_lifecycle_controller.go#L152

Ideas: We could check for deleted and then check for shutdown.  Or we could change `InstanceShutdownByProviderID`  to return an enum (though which values would be choose?)

I'm also not certain why we rely on the taint at all if we are terminating the node ... this feels like there's a race somewhere if it matters.

Maybe put it on the sig-cloudprovider agenda for this afternoon?

**Comment by mcrute** [comment]
@justinsb Ah I see, I think this [continue](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_lifecycle_controller.go#L164) is really the problem, otherwise [this code in the cloud provider](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L4474-L4475) would get hit and the instance would be deleted. It seems valid to just remove that `continue` and everything else should work correctly. WDYT?

**Comment by zhan849** [comment]
@justinsb thanks for reminding, I just saw that commit as well. If we return true when terminated, the behavior will be - node gets added shutdown taint after the state showed up as "terminated", and after the node gets disappeared from cloud provider, it will proceed to delete.

**Comment by mcrute** [comment]
I think the correct behavior here is to taint the node if it's shutdown then fall-through to check if the node still exists in the cloud provider. Terminated nodes will be removed at this point. I don't know the background on why this short-circuits but my guess it it's just efficiency to avoid another describe call to the CP. If that's the case I suggest we remove the `continue` and let it reap nodes even if they're shutdown.

FWIW, the Azure CP is taking the [same approach](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/azure/azure_instances.go#L201) as this PR so they'd be impacted by this as well.

**Comment by mcrute** [comment]
Terminated instances will, after some time period (~2 hours I think), stop being returned in a `DescribeInstances` call. At this point `InstanceShutdownByProviderID` will return `false` and the node will be removed. The more I look at it, I don't think there's anything wrong with this patch.

**Comment by mcrute** [comment]
@justinsb WDYT of pushing this forward given the comments above?

**Comment by zhan849** [comment]
@gnufied I noticed you were also participating in last few changes regarding node shutdown, want your input as well on this. 

*TL;DR:* this fix does not change current behavior that after node is shutdown (stopped or terminated), we keep the `v1.Node` object in api server until the instance disappear from cloud provider, but just make node lifecycle controller to add a shutdown taint to the `v1.Node` object which makes it less confusing to other apps such as daemonset.

**Comment by gnufied** [comment]
Sorry if this was answered before but what problems does arise if we *don't merge* this patch? Is the problem that a terminated node (before 1 hour time) is getting pods scheduled to it by the scheduler?

**Comment by zhan849** [comment]
> Sorry if this was answered before but what problems does arise if we _don't merge_ this patch? Is the problem that a terminated node (before 1 hour time) is getting pods scheduled to it by the scheduler?

@gnufied the problem is that if we don’t call a terminated (not just stopped) node as “shutdown”, the shutdown taint will not be applied to terminated nodes, which will make daemonset controller confused that such node is temporarily unreachable, and schedule a daemonset pod onto terminated nodes.

**Comment by gnufied** [comment]
@zhan849 okay thanks for the explanation. I am curious how this patch affects pods that use EBS volumes.  

As I understand this patch will cause node to stick around until it is really removed from cloudprovider, which will mean volume will not be detached from this node (even though no detach may be necessary), and attach/detach controller will consider EBS volume in-use. This PR may actually worsen time it takes for a pod with EBS volume to start on the replacement node. :( 

I am not sure I like the tradeoff.

**Comment by zhan849** [comment]
@gnufied actually without this patch, both stopped / terminated nodes will stick around for about 1hour, but stopped nodes will have a shutdown taint added from node lifecycle controller but terminated node does not have this taint.

This patch will result in terminated nodes to have a “shutdown” taint as well.

Give this, if attach detach controller does not handle volume state based on the specific shutdown taint on node, sounds like it’s behavior will not change. Can you pls confirm?

**Comment by gnufied** [comment]
Terminated nodes disappear from API server almost immediately. Because even though they are listed by AWS (cloudprovider), `InstanceExistsByProviderID` check below in node_lifecycle_controller.go causes the node object in api-server to be deleted. 

I in-fact do not understand what this patch is trying to fix. If instance is listed by the cloudprovider and has `InstanceStateNameTerminated` set then `ensureNodeExistsByProviderID` check in node_life_cycle_controller.go will cause node object in API server to be deleted.

**Comment by gnufied** [comment]
In this PR - both `shutdownInCloudProvider` and `ensureNodeExistsByProviderID` use same state - `InstanceStateNameTerminated` but my point is, if instance really does have that state then it should be removed from api-server. If it does not have that state, then you can't apply shutdown taint as well.  I do not see what is causing a terminated instance to stick around in api-server and if it does in-fact sticks around then likely `InstanceStateNameTerminated` won't be set as well (in which case this PR won't work). 

What am I missing?

**Comment by gnufied** [comment]
FWIW volume behaviour *will* change because of this PR.  attach/detach controller does not consider node taint for volume detaches, it only relies on pod state (to be deleted). 

What this PR will do is - not removing a terminated node will cause pods on that node to be not deleted too (similar to shutdown nodes) and attach/detach controller will consider those volumes in-use and not perform detaches until pods are deleted. 

Deleting a node from api-server however causes all pods on that node to be deleted and hence volumes are immediately detached.

**Comment by zhan849** [comment]
@gnufied thanks for explaining - I must have slipped the update to `InstanceExistsByProviderID` impl made late last year. So looks like currently we assume "terminated" node does not exist in cloud provider https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L1556 and we proceed to delete terminated instance immediately.

this makes sense to me and I will drop this issue cc @mcrute 

/close
